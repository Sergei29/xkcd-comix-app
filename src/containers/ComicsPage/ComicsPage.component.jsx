import React, { useEffect, useState, useRef } from "react";
import { connect } from "react-redux";
import { fetchComicsById } from "../../redux/actions";
import ComicsView from "../../components/Comics/ComicsView.component";

export const ComicsPage = ({ match, loading, comicsList, fetchComicsById }) => {
	const { id } = match.params;
	const status = useRef({ willUnmount: false });
	const [currentComics, setCurrentComics] = useState({});
	const { title, img, num, alt } = currentComics;

	useEffect(() => {
		if (!status.current.willUnmount) {
			fetchComicsById(id, comicsList);
		}
		if (comicsList[id]) {
			setCurrentComics(comicsList[id]);
		}
	}, [id, comicsList, fetchComicsById]);

	useEffect(() => {
		return () => {
			// cleanup for all api calls before unmount:
			status.current.willUnmount = true;
		};
	}, []);
	return (
		<>
			<h1>Comics page {match.params.id}</h1>
			{loading ? (
				<p>Loading...</p>
			) : (
				<ComicsView title={title} img={img} num={num} alt={alt} />
			)}
		</>
	);
};

const mapStateToProps = (state) => ({
	comicsList: state.comicsList,
	loading: state.loading,
});

const mapDispatchToProps = (dispatch) => ({
	fetchComicsById: (id, comicsList) =>
		dispatch(fetchComicsById(id, comicsList)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ComicsPage);
