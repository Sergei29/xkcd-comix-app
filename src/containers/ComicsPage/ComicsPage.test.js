import { shallow } from "enzyme";
import React from "react";

import { ComicsPage } from "./ComicsPage.component";
import ComicsView from "../../components/Comics/ComicsView.component";

describe("ComicsPage", () => {
	let wrapper;
	const mockProps = {
		loading: false,
		comicsList: {
			614: {
				title: "ABC",
				img: "https:://my-img.png",
				num: 123,
				alt: "BCD",
			},
		},
		fetchComicsById: jest.fn(),
		match: { params: { id: 614 } },
	};

	beforeEach(() => {
		wrapper = shallow(<ComicsPage {...mockProps} />);
	});
	it("should render the component without crashing", () => {
		expect(wrapper.length).toEqual(1);
	});

	it("should contain ComicsView component", () => {
		expect(wrapper.containsMatchingElement(<ComicsView />)).toEqual(true);
	});

	it("should match the snapshot", () => {
		expect(wrapper).toMatchSnapshot();
	});
});
