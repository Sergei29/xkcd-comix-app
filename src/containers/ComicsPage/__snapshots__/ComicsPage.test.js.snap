// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComicsPage should match the snapshot 1`] = `ShallowWrapper {}`;
